Server/index.ts

```diff
@@ -39,103 +72,134 @@ io.on("connection", (socket: Socket) => {
     }
     // This event is for testing only
     socket.on("test", (data)=> {
-        const mkPlayer = () => ({id: 'a', name: 'a', faction: "T" as Faction, team: 0, playerInfo: undefined});
-        const a = new GameInfo([mkPlayer(), mkPlayer(), mkPlayer(), mkPlayer()]);
+        const mkPlayer = () => ({id: 'a', name: 'a', faction: "T" as Faction, team: 0, playerInfo: undefined, status: PlayerStatus.Active});
+        const a = new GameState([mkPlayer(), mkPlayer(), mkPlayer(), mkPlayer()]);
         console.log(a);
     });
     socket.on("disconnect", (reason)=> {
         console.log(socket.id + " has disconnected.");//
         // For now, we don't care about the reason
         const sock = socketTable[socket.id];
         if (sock) {
-            // if (sock.state === SocketState.Menu) // Nothing to do here
-            if (sock.state === SocketState.Lobby) {
-                // Remove player from lobby
-                const {player, game} = sock.info!;
-                game.players = game.players.filter(x => x.id !== socket.id);
-                if (game.players.length === 0) {
-                    // Close lobby if lobby is empty
-                    delete lobbyTable[game.id];
-                } else {
-                    socket.to(game.id).emit("player-left-lobby", sock.clientId);
+            switch(sock.state) {
+                case SocketState.Menu:
+                    break; // Nothing to do here
+                case SocketState.GameEnd:
+                    // Remove player from lobby
+                case SocketState.Lobby: {
+                    // Remove player from lobby
+                    const {player, lobby} = sock.info!;
+                    lobby.players = lobby.players.filter(x => x.id !== socket.id);
+                    if (lobby.players.length === 0) {
+                        // Close lobby if lobby is empty
+                        delete lobbyTable[lobby.id];
+                    } else {
+                        socket.to(lobby.id).emit("player-left-lobby", sock.clientId);
+                    }
                 }
-            } else if (sock.state === SocketState.Game) {
-                // const {player, game} = sock.info as {game: Game, player: Player};
-                // Do other stuff here as appropriate
-                // emit "player-left-game", sock.clientId
-            } // Else, well, something's weird
+                case SocketState.Game: {
+                    // TODO
+                    const {player, lobby} = sock.info;
+                    // Mark status as disconnected
+                    player.status = PlayerStatus.Disconnected;
+                    // End turn automatically
+                    let status = false;
+                    if (lobby.gameInfo!.active) status = lobby.gameInfo!.changeEndStatus(player.playerInfo!.self, true);
+                    if (lobby.players.reduce((acc: number, e) => acc + e.status !== PlayerStatus.Disconnected ? 1 : 0, 0) > 0) {

```

Comment:
```suggestion
                    if (lobby.players.some(player => player.status !== PlayerStatus.Disconnected) {
```
This is simpler and closer to what you really mean. In general, a good test for this is to say in English what you want to write, then convert it into code, then convert the code back into English and compare it with the original. For example:

1: English: is there a player in the lobby that is disconnected?
2: Code: `lobby.players.reduce((acc: number, e) => acc + e.status !== PlayerStatus.Disconnected ? 1 : 0, 0) > 0`
3: Back to English: starting the sum at 0, go through each player in the lobby. If the player is disconnected, add 1 to the sum, otherwise add 0. Is the sum positive?

See how different 1 is from 3?